src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -147,29 +166,50 @@ public static void waitForStart(@NonNull String k, @CheckForNull Run<?,?> b) thr
                 if (b != null && !b.isBuilding()) {
                     throw new AssertionError(JenkinsRule.getLog(b));
                 }
-                s.wait(1000);
+                s.wait(100);
             }
         }
     }
 
+    @Override public StepExecution start(StepContext context) throws Exception {
+        return new Execution(context, k());
+    }
+
     public static class Execution extends AbstractStepExecutionImpl {
 
-        @Inject(optional=true) private SemaphoreStep step;
-        private String k;
+        private final String k;
+
+        Execution(StepContext context, String k) {
+            super(context);
+            this.k = k;
+        }
 
         @Override public boolean start() throws Exception {
             State s = State.get();
-            k = step.k();
-            boolean sync = true;
-            if (s.returnValues.containsKey(k)) {
-                System.err.println("Immediately running " + k);
-                getContext().onSuccess(s.returnValues.get(k));
-            } else if (s.errors.containsKey(k)) {
-                System.err.println("Immediately failing " + k);
-                getContext().onFailure(s.errors.get(k));
+            Object returnValue = null;
+            Throwable error = null;
+            boolean success = false, failure = false, sync = true;
+            synchronized (s) {
```

Comment:
Another point where synchronization was inadequate. Rather cumbersome idiom to avoid holding the lock while making Pipeline API calls, to avoid possible deadlocks.